### PR TITLE
fixes #26627 - rake task to render templates from dir

### PR DESCRIPTION
**Pull request by timogoebel**
This adds a rake task that renders templates from a directory.
This could be used in the [community templates](https://github.com/theforeman/community-templates) CI like this:

```bash
git checkout https://github.com/theforeman/foreman.git
bundle install, ...

bundle exec rake templates:render PATH="path/to/community-templates/provisioning_templates/"
```

This commit also contains fixes so that the templates actually render. We should probably fix them in separate PRs (with proper test coverage).

See [Discourse](https://community.theforeman.org/t/provisioning-templates-testing-separate-repository/13654) for Details.

### CODE REVIEW

File: app/services/foreman/render_templates_from_folder.rb
Comment (ekohl) [inline app/services/foreman/render_templates_from_folder.rb]:
```suggestion
      templates.select(&:snippet?)
```

File: app/services/foreman/render_templates_from_folder.rb
Comment (ekohl) [inline app/services/foreman/render_templates_from_folder.rb]: IMHO we shouldn't assume any OS.
```suggestion
      oses = metadata[:oses] || []
```

File: app/services/foreman/render_templates_from_folder.rb
Comment (ekohl) [inline app/services/foreman/render_templates_from_folder.rb]: There's no way to override or deduce this?

File: app/services/foreman/render_templates_from_folder.rb
Comment (ekohl) [inline app/services/foreman/render_templates_from_folder.rb:104]: What's the `:` in the argument? Is this a ruby syntax I'm unfamiliar with?

File: app/services/foreman/render_templates_from_folder.rb
Comment (ekohl) [inline app/services/foreman/render_templates_from_folder.rb:108]: Again, not a fan of assuming RH. I wouldn't mind a hard error.

File: app/services/foreman/render_templates_from_folder.rb
Comment (ekohl) [inline app/services/foreman/render_templates_from_folder.rb]: We could use the RFC5735 name for it:
```suggestion
        name: 'TEST-NET-1',
```

**Review by ekohl — COMMENTED** [review]
I like this approach a lot. It makes it easy for users to have their custom git repository and render their templates in a CI. RPMs or even containers can probably be used by them.

I'd be happy to rewrite the community-templates testing to use this and then render it with pykickstart or verify it with debian as we do now.

File: app/services/foreman/render_templates_from_folder.rb
Comment (timogoebel) [reply app/services/foreman/render_templates_from_folder.rb:104]: That‘s kwargs in ruby. ;-)

File: app/services/foreman/render_templates_from_folder.rb
Comment (timogoebel) [reply app/services/foreman/render_templates_from_folder.rb:108]: Snippets do not have an OS. So this is ok here imho. And we‘re actually not assuming RedHat here.

**Comment by timogoebel** [comment]
@ekohl: Updated and added some small tests. I think this is ready to be tested within community-templates. ;-)

**Comment by lzap** [comment]
> I like this approach a lot. It makes it easy for users to have their custom git repository and render their templates in a CI. RPMs or even containers can probably be used by them.

I like this too, however one thing comes to my mind as I stated in the discourse. I think that we should move snapshot files to community-templates even at cost that we would keep the test code in core. Because a single PR should always change both templates and snapshot files in a single transaction. This is the biggest value of snapshot tests that we actually see rendered templates during review - it's much easier to verify rendered stuff and this also forces authors to run the snapshot test which easily reveal syntax errors in ERB.

> I'd be happy to rewrite the community-templates testing to use this and then render it with pykickstart or verify it with debian as we do now.

We can even move this stuff here - simply extend the renderer - if it's kickstart file then run pykickstart. Of course in another PR, out of table for now but great idea. I guess we can do this in a generic way and also check YAML etc.

**Comment by timogoebel** [comment]
> I guess we can do this in a generic way and also check YAML etc.

Yeah, that would be really helpful to catch issues early.

> I think that we should move snapshot files to community-templates even at cost that we would keep the test code in core.

That should be fairly easy to do. I'm actually thinking of how we can merge the code that generates the snapshots and this code.

**Comment by lzap** [comment]
> how we can merge the code

What's that? Merge? Not following. :)

**Comment by lzap** [comment]
Oh you mean getting generation working with the directory renderer?

**Comment by timogoebel** [comment]
> Oh you mean getting generation working with the directory renderer?

Yep. :-)

File: lib/tasks/templates_renderer.rake
Comment (lzap) [inline lib/tasks/templates_renderer.rake]: I can live with that but `FOLDER`? Really? ;-) Sounds like a `PATH` to me.

Comment (lzap) [reply lib/tasks/templates_renderer.rake]: Holy crap that was the reason why it could not find `git` on PATH. Huh. Nice.

File: lib/tasks/templates_renderer.rake
Comment (lzap) [inline lib/tasks/templates_renderer.rake:29]: Could you actually print template filepath so we can quickly copy and paste it?

File: test/models/hosts/managed_test.rb
Comment (ekohl) [inline test/models/hosts/managed_test.rb]: Nit: this can be a valid IP. Ideally something from a TEST-NET-* range is used.

**Comment by lzap** [comment]
Looks good, works fine. What is the plan then? Let's merge this and run the rake task from the community-templates repo feeding it with up to date templates? I realized this is just rendering test, no snapshots involved so disregard my previous comment.

File: lib/tasks/templates_renderer.rake
Comment (ares) [reply lib/tasks/templates_renderer.rake]: Folder vs Directory, round 1, fight :-) I don't mind, just wanted to be funny.

Comment (timogoebel) [reply lib/tasks/templates_renderer.rake]: I think I like `DIRECTORY` more. :-)

**Comment by timogoebel** [comment]
> Looks good, works fine. What is the plan then? Let's merge this and run the rake task from the community-templates repo feeding it with up to date templates?

That would be my suggestion. Let's merge this (after I fixed the nits) and integrate it into community-templates CI. We'd benefit from better tests there and can then go further.

File: lib/tasks/templates_renderer.rake
Comment (ekohl) [reply lib/tasks/templates_renderer.rake]: Rake tasks can have arguments, making it self documenting. Would that be a better idea?

Comment (timogoebel) [reply lib/tasks/templates_renderer.rake]: I chose env variables because they have named arguments. If we extend the task to write the rendered templates to files, we'd need a second argument.

But I must admit this looks pretty nice:
```
foreman-rake templates:render['/tmp/community-templates/provisioning_templates']
```

**Comment by lzap** [comment]
What's going on with this PR? I had one comment "Could you actually print template filepath so we can quickly copy and paste it?" and then I was ready to merge, now it looks like you are breaking this up into smaller PRs. That's fine, just let me know when you want to re-test this one.

**Comment by timogoebel** [comment]
> What's going on with this PR?

I want to have the unrelated fixes in separate commits (with proper tests) so they can be cherry-picked. I'll rebase & revisit this patch when the PRs are merged (should be soon I believe).

**Comment by timogoebel** [comment]
@lzap: All done, rebased and added printing the path to the template.

**Comment by timogoebel** [comment]
@lzap: Anything missing here?

**Review by lzap — APPROVED** [review]
Works fine, thanks.

**Comment by ares** [comment]
@lzap please don't forget to set the fixed in version in redmine after merging

**Pull request merged**
